### PR TITLE
feat: Add `pip_tmp_dir` to allow setting the location of the pip temporary directory

### PR DESCRIPTION
**Pull request by artificial-aidan**
…rectory

Signed-off-by: Aidan Jensen <aidan@artificial.com>

## Description

Add the ability to set the location of the temporary directory for pip installs

## Motivation and Context

When building in remote docker, you may need to set a shared location for the pip install directory

## Breaking Changes

No it does not

## How Has This Been Tested?
- [X] I have tested and validated these changes using one or more of the provided `examples/*` projects

I added a test that uses a temp dir in the cwd, and watched that directory while it was used.

### CODE REVIEW

**Comment by artificial-aidan** [comment]
@antonbabenko any thoughts on this?

**Comment by antonbabenko** [comment]
@pdecat Could you please review this one (especially the Python part)?

File: README.md [hunk 1]
```diff
@@ -599,10 +601,10 @@ Q4: What does this error mean - `"We currently do not support adding policies fo
 
 | Name | Version |
 |------|---------|
-| <a name="provider_aws"></a> [aws](#provider\_aws) | >= 3.61 |
-| <a name="provider_external"></a> [external](#provider\_external) | >= 1 |
-| <a name="provider_local"></a> [local](#provider\_local) | >= 1 |
-| <a name="provider_null"></a> [null](#provider\_null) | >= 2 |
+| <a name="provider_aws"></a> [aws](#provider\_aws) | 3.64.2 |
```
Comment (pdecat) [inline README.md:604]: I believe all these changes not related to the addition of `pip_tmp_dir` should not be included in this PR.

@antonbabenko that's an issue I've always faced when running the terraform-docs pre-commit hook in this repository, but never asked about: what are we supposed to do about that?

File: examples/build-package/main.tf [hunk 2]
```diff
@@ -27,6 +27,20 @@ module "package_dir" {
...
+# Create zip-archive of a single directory where "pip install" will also be executed (default for python runtime) and set temporary directory for pip install
+module "package_dir_pip_dir" {
+  source = "../../"
+
+  create_function = false
+
+  runtime = "python3.8"
+  source_path = [{
+    path             = "${path.module}/../fixtures/python3.8-app1"
+    pip_tmp_dir      = "${path.cwd}/../fixtures"
```
Comment (pdecat) [inline examples/build-package/main.tf:39]: Why `path.cwd` instead of `path.module`?

File: package.py [hunk 3]
```diff
@@ -836,6 +836,7 @@ def install_pip_requirements(query, requirements_file):
     runtime = query.runtime
     artifacts_dir = query.artifacts_dir
     docker = query.docker
+    temp_dir = query.temp_dir
```
Comment (pdecat) [inline package.py:839]: Isn't this `temp_dir` shadowed by `with tempdir(tmp_dir) as temp_dir` at line [872](https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/230/files#diff-34b2a63c4b18a635042766329eca28d8801699eecf5e93eaedaeda1309931381L872)?

File: package.py [hunk 4]
```diff
@@ -825,7 +825,7 @@ def _zip_write_with_filter(zip_stream, path_filter, source_path, prefix,
 
 
 @contextmanager
-def install_pip_requirements(query, requirements_file):
+def install_pip_requirements(query, requirements_file, tmp_dir=None):
```
Comment (pdecat) [inline package.py:828]: I believe having a default `None` value is not needed as `install_pip_requirements()` is [called from a single place.](https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/230/files#diff-34b2a63c4b18a635042766329eca28d8801699eecf5e93eaedaeda1309931381L788)

**Review by pdecat — COMMENTED** [review]
Hi, most changes relative to the addition of `pip_tmp_dir` look correct to me, left a few remarks.

However, I'm still wondering about a concrete need for that.

@artificial-aidan, you mention "_Can be useful for Docker in Docker builds._". Can you expand a bit why? In your actual use case, do you specify a special path that can be cached or something?

Also, wouldn't [setting `TMPDIR`, `TEMP` or `TMP` environment variable](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp) achieve the same?

**Comment by artificial-aidan** [comment]
> Hi, most changes relative to the addition of `pip_tmp_dir` look correct to me, left a few remarks.
> 
> However, I'm still wondering about a concrete need for that.
> 
> @artificial-aidan, you mention "_Can be useful for Docker in Docker builds._". Can you expand a bit why? In your actual use case, do you specify a special path that can be cached or something?
> 
> Also, wouldn't [setting `TMPDIR`, `TEMP` or `TMP` environment variable](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp) achieve the same?

When using a remote Docker container for docker builds, the `tmp` directory might not be mounted on the remote host. In my case there is a shared directory mounted between the caller and the remote docker container. By setting the temp dir to this, the docker builder can access the files created.

In theory those environment variables might work, (and would have saved me a lot of time probably) but now that I wrote it it seems like there could be a case where you didn't want all of the python temporary dirs to go to wherever you are setting your pip_temp_dir to be.

File: package.py (hunk 4 above)
Comment (artificial-aidan) [reply package.py:828]: I was just being safe, but I can remove if needed.

File: README.md (hunk 1 above)
Comment (artificial-aidan) [reply README.md:604]: I was following the instructions on another PR. But I also don't know what to do with this one. It happened just when running the hooks.

File: examples/build-package/main.tf (hunk 2 above)
Comment (artificial-aidan) [reply examples/build-package/main.tf:39]: Trying to remember now, but I believe that `path.module` has a `../` prefix in it. Which something down the line did not like.

File: package.py (hunk 3 above)
Comment (artificial-aidan) [reply package.py:839]: I believe this may have been left over from a first pass I did. I will remove.

Comment (artificial-aidan) [reply package.py:839]:
```suggestion
```

File: README.md (hunk 1 above)
Comment (pdecat) [reply README.md:604]: What seems especially wrong to me is replacing the flexible constraints coming from `versions.tf` by hard-coded version numbers. I believe this is caused by the presence of `.terraform.lock.hcl` files.

Probably caused by the fact that we are using too recent (>= 0.14) versions of terraform-docs that include https://github.com/terraform-docs/terraform-docs/pull/509, while the [Github Actions CI is using v0.13.0](https://github.com/terraform-aws-modules/terraform-aws-lambda/blob/v2.26.0/.github/workflows/pre-commit.yml#L94).

Note: terraform-docs v0.15 introduced a `--lockfile` parameters that allows to disable that behavior: https://github.com/terraform-docs/terraform-docs/pull/527

Comment (antonbabenko) [reply README.md:604]: Skip this issue, please. I know it was problematic for many for some time. I will fix the docs to make checks green if necessary after this PR is ready.

We are going to fix it in the near future like it is done here - https://github.com/terraform-aws-modules/terraform-aws-rds/pull/361

File: examples/build-package/main.tf (hunk 2 above)
Comment (antonbabenko) [reply examples/build-package/main.tf:39]: `path.module` seems to be a better choice here ([see docs about `path.cwd`](https://www.terraform.io/docs/language/expressions/references.html#path-cwd)).

Optionally, user can use something like - `abspath(path.module)` if absolute path is required.

File: package.py (hunk 4 above)
Comment (antonbabenko) [reply package.py:828]: @artificial-aidan Please do.

File: examples/build-package/main.tf [hunk 5]
```diff
@@ -27,6 +27,20 @@ module "package_dir" {
   source_path = "${path.module}/../fixtures/python3.8-app1"
 }
 
+# Create zip-archive of a single directory where "pip install" will also be executed (default for python runtime) and set temporary directory for pip install
```
Comment (antonbabenko) [inline examples/build-package/main.tf:30]: Thank you for making an example for this feature!

**Review by antonbabenko — COMMENTED** [review]
Just minor comments. Leave terraform-docs as the last step (I will fix).

**Comment by artificial-aidan** [comment]
> Just minor comments. Leave terraform-docs as the last step (I will fix).

Wrapped up the minor comments and fixed a merge conflict.

**Review by ahlinc — APPROVED** [review]

**Comment by antonbabenko** [comment]
This PR is included in version 2.28.0 :tada:
